### ISSUE TITLE
Optimize has() function 

### ISSUE THREAD

**Issue by coolguru** [dgraph-io/dgraph#2671]
### What you wanted to do
We're using a strategy of giving types to nodes recommended in this link https://docs.dgraph.io/howto/#giving-nodes-a-type. In our queries, we're using an indexed field in root func first and then using filter to look for a particular type. has function is actually loading all the nodes with requested type instead of just filtering on nodes that matches the root function criteria. 
`
query 
{
  q(func:eq(first, "Paul")) @filter(has(CTO)) {
   expand(_all_) {
    uid
  }
  }
}`

`
//code to generate test data
package main

import (
	"context"
	"encoding/json"
	"flag"
	"fmt"
	"math/rand"
	"strings"
	"sync/atomic"
	"time"

	"github.com/dgraph-io/dgo"
	"github.com/dgraph-io/dgo/protos/api"
	"github.com/dgraph-io/dgo/x"
	"github.com/dgraph-io/dgo/y"
	"google.golang.org/grpc"
)

var (
	addr    = flag.String("addr", "localhost:9080", "dgraph address")
	concurr = flag.Int("c", 3, "number of concurrent upserts per account")
)

var (
	firsts = []string{"Paul", "Eric", "Jack", "John", "Martin"}
	lasts  = []string{"Brown", "Smith", "Robinson", "Waters", "Taylor"}
	types  = []string{"CEO", "COO", "CTO", "CFO"}
)

type account struct {
	first string
	last  string
	age   int
}

var accounts []account

func init() {
	for _, first := range firsts {
		for _, last := range lasts {
			for i := 0; i < 10000; i++ {
				accounts = append(accounts, account{
					first: first,
					last:  last,
					age:   i,
				})
			}
		}
	}
}

func main() {
	flag.Parse()
	c := newClient()
	setup(c)
	fmt.Println("Doing upserts")
	doUpserts(c)
	fmt.Println("Checking integrity")
	checkIntegrity(c)
}

func newClient() *dgo.Dgraph {
	d, err := grpc.Dial(*addr, grpc.WithInsecure())
	x.Check(err)
	return dgo.NewDgraphClient(
		api.NewDgraphClient(d),
	)
}

func setup(c *dgo.Dgraph) {
	ctx := context.Background()
	x.Check(c.Alter(ctx, &api.Operation{
		DropAll: true,
	}))
	x.Check(c.Alter(ctx, &api.Operation{
		Schema: `
			first:  string   @index(term) @upsert .
			last:   string   @index(hash) @upsert .
			age:    int      @index(int)  @upsert .
			when:   int                   .
		`,
	}))
}

func doUpserts(c *dgo.Dgraph) {
	for _, acct := range accounts {
		upsert(c, acct)
	}
}

var (
	successCount uint64
	retryCount   uint64
	lastStatus   time.Time
)

func upsert(c *dgo.Dgraph, acc account) {
	for {
		if time.Since(lastStatus) > 100*time.Millisecond {
			fmt.Printf("[%s] Success: %d Retries: %d\n", time.Now().Format(time.Stamp),
				atomic.LoadUint64(&successCount), atomic.LoadUint64(&retryCount))
			lastStatus = time.Now()
		}
		err := tryUpsert(c, acc)
		if err == nil {
			atomic.AddUint64(&successCount, 1)
			return
		} else if err == y.ErrAborted {
			// pass
		} else {
			fmt.Errorf("ERROR: %v", err)
		}
		atomic.AddUint64(&retryCount, 1)
	}
}

func tryUpsert(c *dgo.Dgraph, acc account) error {
	ctx := context.Background()

	txn := c.NewTxn()
	defer txn.Discard(ctx)
	q := fmt.Sprintf(`
		{
			get(func: eq(first, %q)) @filter(eq(last, %q) AND eq(age, %d)) {
				uid expand(_all_) {uid}
			}
		}
	`, acc.first, acc.last, acc.age)
	resp, err := txn.Query(ctx, q)
	x.Check(err)

	decode := struct {
		Get []struct {
			Uid *string
		}
	}{}
	x.Check(json.Unmarshal(resp.GetJson(), &decode))

	x.AssertTrue(len(decode.Get) <= 1)
	s := rand.NewSource(time.Now().Unix())
	r := rand.New(s) // initialize local pseudorandom generator
	t := r.Intn(len(types))

	var uid string
	if len(decode.Get) == 1 {
		x.AssertTrue(decode.Get[0].Uid != nil)
		uid = *decode.Get[0].Uid
	} else {
		nqs := fmt.Sprintf(`
			_:acct <first> %q .
			_:acct <last>  %q .
			_:acct <age>   "%d"^^<xs:int> .
			_:acct <type> %q .
			_:acct <%s> "" .
		`,
			acc.first, acc.last, acc.age, types[t], types[t],
		)
		mu := &api.Mutation{SetNquads: []byte(nqs)}
		assigned, err := txn.Mutate(ctx, mu)
		if err != nil {
			return err
		}
		uid = assigned.GetUids()["acct"]
		x.AssertTrue(uid != "")

	}

	nq := fmt.Sprintf(`
		<%s> <when> "%d"^^<xs:int> .
	`,
		uid, time.Now().Nanosecond(),
	)
	mu := &api.Mutation{SetNquads: []byte(nq)}
	if _, err = txn.Mutate(ctx, mu); err != nil {
		return err
	}

	return txn.Commit(ctx)
}

func checkIntegrity(c *dgo.Dgraph) {
	ctx := context.Background()

	q := fmt.Sprintf(`
		{
			all(func: anyofterms(first, %q)) {
				first
				last
				age
			}
		}
	`, strings.Join(firsts, " "))
	resp, err := c.NewTxn().Query(ctx, q)
	x.Check(err)

	decode := struct {
		All []struct {
			First *string
			Last  *string
			Age   *int
		}
	}{}
	x.Check(json.Unmarshal(resp.GetJson(), &decode))

	// Make sure there is exactly one of each account.
	accountSet := make(map[string]struct{})
	for _, record := range decode.All {
		x.AssertTrue(record.First != nil)
		x.AssertTrue(record.Last != nil)
		x.AssertTrue(record.Age != nil)
		str := fmt.Sprintf("%s_%s_%d", *record.First, *record.Last, *record.Age)
		accountSet[str] = struct{}{}
	}
	x.AssertTrue(len(accountSet) == len(accounts))
	for _, acct := range accounts {
		str := fmt.Sprintf("%s_%s_%d", acct.first, acct.last, acct.age)
		_, ok := accountSet[str]
		x.AssertTrue(ok)
	}
}
`

**Comment by makitka2007** [comment]
oh my god, is it true? we have a lot of "has" filters in queries, i was sure that it will work as filter.
mmm, if i have "regexp" filter - will it retrieve all nodes as well instead of filtering already returned ones?

**Comment by makitka2007** [comment]
btw, i tested 2 queries on 10 millions edges and found that
`{ ddd(func: has(type_Work)) { count(uid) } }` takes more than 2 minutes
whereas `{ ddd(func: eq(entity_key, "entity7777")) @filter(has(type_Work)) { count(uid) } }` works very fast (tens milliseconds).
so, seems like "has" function in filter really works as filter. i didn't dig into dgraph code, just made some conclusions based on my tests

**Comment by makitka2007** [comment]
i don’t think so, i tried on different keys and if was always fast, and first query was always much more slower even when i retried it. but i dont’t know know how it work internally and i hope developers team will shed light on it

**Comment by makitka2007** [comment]
and even this one is running slow too - `{ ddd(func: has(type_Work)) @filter(eq(entity_key, "entity888")) { count(uid) } }`, just to return 1 node - 144 seconds comparing to `{ ddd(func: eq(entity_key, "entity888")) @filter(has(type_Work)) { count(uid) } }` which takes 6ms only
